File: server/routes/projects.js
const express = require("express");
const { db } = require("../db");
const router = express.Router();

router.get("/", (req, res) => {
  const projects = db.prepare("SELECT * FROM projects").all();
  res.json(projects);
});

router.get("/:id", (req, res) => {
  const project = db
    .prepare("SELECT * FROM projects WHERE id = ?")
    .get(req.params.id);
  if (!project) return res.status(404).json({ error: "Project not found" });

  const experiments = db
    .prepare("SELECT * FROM experiments WHERE iacpj_nm = ?")
    .all(project.iacpj_nm);
  res.json({ ...project, experiments });
});

// 과제 추가
router.post("/", (req, res) => {
  const { iacpj_nm } = req.body;
  if (!iacpj_nm || !iacpj_nm.trim()) {
    return res.status(400).json({ error: "iacpj_nm은 필수입니다." });
  }

  try {
    const stmt = db.prepare(`
      INSERT INTO projects (
        iacpj_nm, iacpj_tgt_n, iacpj_level, iacpj_tech_n,
        ia_tgt_htr_n, iacpj_nud_n, iacpj_mod_n, iacpj_itf_uno, iacpj_bgn_dy,
        iacpj_ch_n, ia_ta_grd_n, project_purpose, iacpj_ta_goa, iacpj_cur_stt,
        iacpj_ch_i, ia_ch_or_i, ia_ch_or_n, ia_ch_or_path, iacpj_core_tec,
        iacpj_end_dy, iacpj_reg_dy
      ) VALUES (
        @iacpj_nm, @iacpj_tgt_n, @iacpj_level, @iacpj_tech_n,
        @ia_tgt_htr_n, @iacpj_nud_n, @iacpj_mod_n, @iacpj_itf_uno, @iacpj_bgn_dy,
        @iacpj_ch_n, @ia_ta_grd_n, @project_purpose, @iacpj_ta_goa, @iacpj_cur_stt,
        @iacpj_ch_i, @ia_ch_or_i, @ia_ch_or_n, @ia_ch_or_path, @iacpj_core_tec,
        @iacpj_end_dy, @iacpj_reg_dy
      )
    `);

    const params = {
      iacpj_nm: req.body.iacpj_nm?.trim(),
      iacpj_tgt_n: req.body.iacpj_tgt_n || null,
      iacpj_level: req.body.iacpj_level || null,
      iacpj_tech_n: req.body.iacpj_tech_n || null,
      ia_tgt_htr_n: req.body.ia_tgt_htr_n || null,
      iacpj_nud_n: req.body.iacpj_nud_n || null,
      iacpj_mod_n: req.body.iacpj_mod_n || null,
      iacpj_itf_uno: req.body.iacpj_itf_uno || null,
      iacpj_bgn_dy: req.body.iacpj_bgn_dy || null,
      iacpj_ch_n: req.body.iacpj_ch_n || null,
      ia_ta_grd_n: req.body.ia_ta_grd_n || null,
      project_purpose: req.body.project_purpose || null,
      iacpj_ta_goa: req.body.iacpj_ta_goa || null,
      iacpj_cur_stt: req.body.iacpj_cur_stt || null,
      iacpj_ch_i: req.body.iacpj_ch_i || null,
      ia_ch_or_i: req.body.ia_ch_or_i || null,
      ia_ch_or_n: req.body.ia_ch_or_n || null,
      ia_ch_or_path: req.body.ia_ch_or_path || null,
      iacpj_core_tec: req.body.iacpj_core_tec || null,
      iacpj_end_dy: req.body.iacpj_end_dy || null,
      iacpj_reg_dy: req.body.iacpj_reg_dy || null,
    };

    const result = stmt.run(params);
    const created = db
      .prepare("SELECT * FROM projects WHERE id = ?")
      .get(result.lastInsertRowid);
    res.status(201).json(created);
  } catch (err) {
    if (err.message.includes("UNIQUE")) {
      return res
        .status(409)
        .json({ error: "이미 동일한 과제명이 존재합니다." });
    }
    console.error("Error creating project:", err);
    res.status(500).json({ error: "과제 생성 중 오류 발생" });
  }
});

// 과제 삭제 (관련 실험, 스플릿도 함께 삭제)
router.delete("/:id", (req, res) => {
  const project = db
    .prepare("SELECT * FROM projects WHERE id = ?")
    .get(req.params.id);
  if (!project) return res.status(404).json({ error: "Project not found" });

  try {
    let counts;
    db.exec("BEGIN");
    try {
      const experiments = db
        .prepare("SELECT plan_id FROM experiments WHERE iacpj_nm = ?")
        .all(project.iacpj_nm);
      for (const exp of experiments) {
        db.prepare("DELETE FROM split_tables WHERE plan_id = ?").run(
          exp.plan_id,
        );
      }
      const expResult = db
        .prepare("DELETE FROM experiments WHERE iacpj_nm = ?")
        .run(project.iacpj_nm);
      db.prepare("DELETE FROM projects WHERE id = ?").run(req.params.id);
      counts = { experiments: expResult.changes, splits: experiments.length };
      db.exec("COMMIT");
    } catch (txErr) {
      db.exec("ROLLBACK");
      throw txErr;
    }
    res.json({
      message: "과제 삭제 완료",
      deleted: { project: project.iacpj_nm, ...counts },
    });
  } catch (err) {
    console.error("Error deleting project:", err);
    res.status(500).json({ error: "과제 삭제 중 오류 발생" });
  }
});

module.exports = router;

File: server_py/routes/projects.py
"""
/api/projects — exact port of server/routes/projects.js
"""
import sqlite3
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse
from ..database import get_db, dict_row, dict_rows

router = APIRouter(prefix="/api/projects", tags=["projects"])

PROJECT_COLS = [
    "iacpj_nm", "iacpj_tgt_n", "iacpj_level", "iacpj_tech_n",
    "ia_tgt_htr_n", "iacpj_nud_n", "iacpj_mod_n", "iacpj_itf_uno", "iacpj_bgn_dy",
    "iacpj_ch_n", "ia_ta_grd_n", "project_purpose", "iacpj_ta_goa", "iacpj_cur_stt",
    "iacpj_ch_i", "ia_ch_or_i", "ia_ch_or_n", "ia_ch_or_path", "iacpj_core_tec",
    "iacpj_end_dy", "iacpj_reg_dy",
]


@router.get("/")
def list_projects(conn: sqlite3.Connection = Depends(get_db)):
    rows = conn.execute("""
        SELECT p.*,
          COALESCE(e.experiment_count, 0) AS experiment_count,
          COALESCE(s.split_count, 0) AS split_count
        FROM projects p
        LEFT JOIN (
          SELECT iacpj_nm, COUNT(*) AS experiment_count
          FROM experiments
          GROUP BY iacpj_nm
        ) e ON p.iacpj_nm = e.iacpj_nm
        LEFT JOIN (
          SELECT ex.iacpj_nm, COUNT(*) AS split_count
          FROM (
            SELECT st.plan_id
            FROM split_tables st
            GROUP BY st.plan_id
            HAVING COUNT(*) >= 2
          ) valid_splits
          JOIN experiments ex ON valid_splits.plan_id = ex.plan_id
          GROUP BY ex.iacpj_nm
        ) s ON p.iacpj_nm = s.iacpj_nm
    """).fetchall()
    return dict_rows(rows)


@router.get("/{project_id}")
def get_project(project_id: int, conn: sqlite3.Connection = Depends(get_db)):
    project = conn.execute("SELECT * FROM projects WHERE id = ?", (project_id,)).fetchone()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    experiments = conn.execute(
        "SELECT * FROM experiments WHERE iacpj_nm = ?", (project["iacpj_nm"],)
    ).fetchall()
    result = dict_row(project)
    result["experiments"] = dict_rows(experiments)
    return result


@router.post("/", status_code=201)
def create_project(body: dict, conn: sqlite3.Connection = Depends(get_db)):
    iacpj_nm = (body.get("iacpj_nm") or "").strip()
    if not iacpj_nm:
        raise HTTPException(status_code=400, detail="iacpj_nm은 필수입니다.")

    params = {col: (body.get(col) or "").strip() or None for col in PROJECT_COLS}
    params["iacpj_nm"] = iacpj_nm

    placeholders = ", ".join(f":{col}" for col in PROJECT_COLS)
    col_names = ", ".join(PROJECT_COLS)

    try:
        cursor = conn.execute(
            f"INSERT INTO projects ({col_names}) VALUES ({placeholders})", params
        )
        conn.commit()
        created = conn.execute(
            "SELECT * FROM projects WHERE id = ?", (cursor.lastrowid,)
        ).fetchone()
        return dict_row(created)
    except sqlite3.IntegrityError as e:
        if "UNIQUE" in str(e):
            raise HTTPException(status_code=409, detail="이미 동일한 과제명이 존재합니다.")
        raise HTTPException(status_code=500, detail="과제 생성 중 오류 발생")


@router.delete("/{project_id}")
def delete_project(project_id: int, conn: sqlite3.Connection = Depends(get_db)):
    project = conn.execute("SELECT * FROM projects WHERE id = ?", (project_id,)).fetchone()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    try:
        experiments = conn.execute(
            "SELECT plan_id FROM experiments WHERE iacpj_nm = ?", (project["iacpj_nm"],)
        ).fetchall()
        for exp in experiments:
            conn.execute("DELETE FROM split_tables WHERE plan_id = ?", (exp["plan_id"],))
        exp_result = conn.execute(
            "DELETE FROM experiments WHERE iacpj_nm = ?", (project["iacpj_nm"],)
        )
        conn.execute("DELETE FROM projects WHERE id = ?", (project_id,))
        conn.commit()
        return {
            "message": "과제 삭제 완료",
            "deleted": {
                "project": project["iacpj_nm"],
                "experiments": exp_result.rowcount,
                "splits": len(experiments),
            },
        }
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail="과제 삭제 중 오류 발생")
